File: lib/pricing-engine/risk.ts
import { Cashflow, NSSParameters } from "./types";
import { ZeroCurve, bootstrapZeroCurve } from "./bootstrap";
import { calculateDirtyPrice } from "./pricing";
import { nssYield, applyScenarioShocks } from "./scenario";

export function calculateMacaulayDuration(
  settlementDate: Date,
  cashflows: Cashflow[],
  ytm: number,
  couponFrequency: number = 2
): number {
  const futureCfs = cashflows.filter(
    cf => (cf.date.getTime() - settlementDate.getTime()) > 0
  );
  if (futureCfs.length === 0) {
    return 0.0;
  }

  let numerator = 0.0;
  let denominator = 0.0;

  for (const cf of futureCfs) {
    const diffDays = Math.round((cf.date.getTime() - settlementDate.getTime()) / (1000 * 60 * 60 * 24));
    const t = diffDays / 365.0;
    const df = 1.0 / Math.pow(1.0 + ytm / (100.0 * couponFrequency), couponFrequency * t);
    const pv = cf.amount * df;
    
    numerator += t * pv;
    denominator += pv;
  }

  if (denominator === 0) {
    return 0.0;
  }
  return numerator / denominator;
}

export function calculateModifiedDuration(
  macaulayDuration: number,
  ytm: number,
  couponFrequency: number = 2
): number {
  return macaulayDuration / (1.0 + ytm / (100.0 * couponFrequency));
}

export function calculateDv01(
  settlementDate: Date,
  cashflows: Cashflow[],
  zc: ZeroCurve
): number {
  const p0 = calculateDirtyPrice(settlementDate, cashflows, zc);
  
  // Parallel shift zero rates up by 1bp (0.01%)
  const bumpedRates = zc.zeroRates.map(r => r + 0.01);
  const zcBumped = new ZeroCurve(zc.maturities, bumpedRates);
  
  const pUp = calculateDirtyPrice(settlementDate, cashflows, zcBumped);
  
  return p0 - pUp;
}

export function calculateConvexity(
  settlementDate: Date,
  cashflows: Cashflow[],
  zc: ZeroCurve
): number {
  const p0 = calculateDirtyPrice(settlementDate, cashflows, zc);
  if (p0 <= 0) {
    return 0.0;
  }

  const shift = 0.1; // 10 bps zero rate shift
  const h = 0.001;   // 10 bps decimal shift
  
  const bumpedRatesUp = zc.zeroRates.map(r => r + shift);
  const zcUp = new ZeroCurve(zc.maturities, bumpedRatesUp);
  const pUp = calculateDirtyPrice(settlementDate, cashflows, zcUp);
  
  const bumpedRatesDown = zc.zeroRates.map(r => r - shift);
  const zcDown = new ZeroCurve(zc.maturities, bumpedRatesDown);
  const pDown = calculateDirtyPrice(settlementDate, cashflows, zcDown);
  
  return (pUp + pDown - 2.0 * p0) / (p0 * (h * h));
}

export interface FactorPnLDecomposition {
  level: number;
  slope: number;
  curvature1: number;
  curvature2: number;
  residual: number;
  total: number;
}

export function calculatePositionFactorPnLDecomposition(
  settlementDate: Date,
  cashflows: Cashflow[],
  baseParams: NSSParameters,
  shocks: {
    parallel_shift?: number;
    slope_shock?: number;
    curvature1_shock?: number;
    curvature2_shock?: number;
    twist_shock?: number;
    twist_pivot?: number;
  },
  faceValue: number = 100.0
): FactorPnLDecomposition {
  const parallelShift = shocks.parallel_shift ?? 0.0;
  const slopeShock = shocks.slope_shock ?? 0.0;
  const curvature1Shock = shocks.curvature1_shock ?? 0.0;
  const curvature2Shock = shocks.curvature2_shock ?? 0.0;
  const twistShock = shocks.twist_shock ?? 0.0;
  const twistPivot = shocks.twist_pivot ?? 5.0;

  const tau1 = baseParams.tau1;
  let g1Pivot = 0.0;
  if (twistPivot > 0) {
    g1Pivot = (1.0 - Math.exp(-twistPivot / tau1)) / (twistPivot / tau1);
  } else {
    g1Pivot = 1.0;
  }

  const deltaBeta0Twist = -twistShock * g1Pivot;
  const deltaBeta0 = parallelShift + deltaBeta0Twist;
  const deltaBeta1 = slopeShock + twistShock;
  const deltaBeta2 = curvature1Shock;
  const deltaBeta3 = curvature2Shock;

  function getPriceForParams(b0: number, b1: number, b2: number, b3: number): number {
    const parCurveFn = (t: number) => {
      return nssYield(t, b0, b1, b2, b3, baseParams.tau1, baseParams.tau2);
    };
    const zc = bootstrapZeroCurve(parCurveFn, 40.0, 0.5);
    return calculateDirtyPrice(settlementDate, cashflows, zc);
  }

  const pBase = getPriceForParams(
    baseParams.beta0,
    baseParams.beta1,
    baseParams.beta2,
    baseParams.beta3
  );

  const shockedParams = applyScenarioShocks(baseParams, shocks);
  const pShocked = getPriceForParams(
    shockedParams.beta0,
    shockedParams.beta1,
    shockedParams.beta2,
    shockedParams.beta3
  );
  const totalPnL = (pShocked - pBase) * (faceValue / 100.0);

  // Central difference bump h = 1bp (0.01 percentage point)
  const h = 0.01;

  const pB0Up = getPriceForParams(baseParams.beta0 + h, baseParams.beta1, baseParams.beta2, baseParams.beta3);
  const pB0Down = getPriceForParams(baseParams.beta0 - h, baseParams.beta1, baseParams.beta2, baseParams.beta3);
  const dB0 = (pB0Up - pB0Down) / (2.0 * h);

  const pB1Up = getPriceForParams(baseParams.beta0, baseParams.beta1 + h, baseParams.beta2, baseParams.beta3);
  const pB1Down = getPriceForParams(baseParams.beta0, baseParams.beta1 - h, baseParams.beta2, baseParams.beta3);
  const dB1 = (pB1Up - pB1Down) / (2.0 * h);

  const pB2Up = getPriceForParams(baseParams.beta0, baseParams.beta1, baseParams.beta2 + h, baseParams.beta3);
  const pB2Down = getPriceForParams(baseParams.beta0, baseParams.beta1, baseParams.beta2 - h, baseParams.beta3);
  const dB2 = (pB2Up - pB2Down) / (2.0 * h);

  const pB3Up = getPriceForParams(baseParams.beta0, baseParams.beta1, baseParams.beta2, baseParams.beta3 + h);
  const pB3Down = getPriceForParams(baseParams.beta0, baseParams.beta1, baseParams.beta2, baseParams.beta3 - h);
  const dB3 = (pB3Up - pB3Down) / (2.0 * h);

  const contribLevel = dB0 * deltaBeta0 * (faceValue / 100.0);
  const contribSlope = dB1 * deltaBeta1 * (faceValue / 100.0);
  const contribCurv1 = dB2 * deltaBeta2 * (faceValue / 100.0);
  const contribCurv2 = dB3 * deltaBeta3 * (faceValue / 100.0);

  const residual = totalPnL - (contribLevel + contribSlope + contribCurv1 + contribCurv2);

  return {
    level: contribLevel,
    slope: contribSlope,
    curvature1: contribCurv1,
    curvature2: contribCurv2,
    residual: residual,
    total: totalPnL
  };
}

File: backend/quant_core/risk.py
from datetime import date
from quant_core.bootstrap import ZeroCurve
from quant_core.pricing import calculate_dirty_price

def calculate_macaulay_duration(
    settlement_date: date,
    cashflows: list,
    ytm: float,
    coupon_frequency: int = 2
) -> float:
    """
    Calculates the Macaulay duration of a bond's future cashflows.
    ytm is in percentage (e.g., 7.28).
    """
    future_cfs = [cf for cf in cashflows if (cf["date"] - settlement_date).days > 0]
    if not future_cfs:
        return 0.0
        
    numerator = 0.0
    denominator = 0.0
    
    for cf in future_cfs:
        t = (cf["date"] - settlement_date).days / 365.0
        # Present value of this cashflow discounted at the YTM rate
        df = 1.0 / ((1.0 + ytm / (100.0 * coupon_frequency)) ** (coupon_frequency * t))
        pv = cf["amount"] * df
        numerator += t * pv
        denominator += pv
        
    if denominator == 0:
        return 0.0
    return numerator / denominator

def calculate_modified_duration(
    macaulay_duration: float,
    ytm: float,
    coupon_frequency: int = 2
) -> float:
    """
    Calculates the Modified duration of a bond.
    ytm is in percentage (e.g., 7.28).
    """
    return macaulay_duration / (1.0 + ytm / (100.0 * coupon_frequency))

def calculate_dv01(settlement_date: date, cashflows: list, zc: ZeroCurve) -> float:
    """
    Calculates the DV01 (dollar value of 01) sensitivity of a bond's price.
    Calculated via a 1 basis point (0.01% parallel increase) bump-and-reprice on the zero curve.
    """
    p0 = calculate_dirty_price(settlement_date, cashflows, zc)
    
    # Parallel shift zero curve up by 1bp (0.01 percentage points)
    bumped_rates = zc.zero_rates + 0.01
    zc_bumped = ZeroCurve(zc.maturities, bumped_rates)
    
    p_up = calculate_dirty_price(settlement_date, cashflows, zc_bumped)
    
    # Price drop for yield increase
    return p0 - p_up

def calculate_convexity(settlement_date: date, cashflows: list, zc: ZeroCurve) -> float:
    """
    Calculates the Convexity of a bond.
    Calculated via a symmetric 10 basis point (0.1% parallel shift up and down) bump-and-reprice on the zero curve.
    Formula: (P_up + P_down - 2 * P_0) / (P_0 * h^2) where h = 0.001 (10 bps in decimal).
    """
    p0 = calculate_dirty_price(settlement_date, cashflows, zc)
    if p0 <= 0:
        return 0.0
        
    shift = 0.1   # 10 bps shift in percentage points
    h = 0.001     # 10 bps shift in decimal yield (0.1 / 100)
    
    zc_up = ZeroCurve(zc.maturities, zc.zero_rates + shift)
    p_up = calculate_dirty_price(settlement_date, cashflows, zc_up)
    
    zc_down = ZeroCurve(zc.maturities, zc.zero_rates - shift)
    p_down = calculate_dirty_price(settlement_date, cashflows, zc_down)
    
    return (p_up + p_down - 2.0 * p0) / (p0 * (h ** 2))

def calculate_position_factor_pnl_decomposition(
    settlement_date: date,
    cashflows: list,
    base_params: dict,
    parallel_shift: float = 0.0,
    slope_shock: float = 0.0,
    curvature1_shock: float = 0.0,
    curvature2_shock: float = 0.0,
    twist_shock: float = 0.0,
    twist_pivot: float = 5.0,
    face_value: float = 100.0,
) -> dict:
    """
    Decomposes the position P&L into contributions from Level, Slope, Curvature-1, and Curvature-2.
    Also returns the joint P&L and the interaction residual. Uses first-order linear attribution
    via central finite differences (h = 1bp).
    """
    import numpy as np
    from quant_core.nss import nss_yield
    from quant_core.bootstrap import bootstrap_zero_curve
    from quant_core.scenario import apply_scenario_shocks

    tau1 = base_params["tau1"]
    if twist_pivot > 0:
        g1_pivot = (1.0 - np.exp(-twist_pivot / tau1)) / (twist_pivot / tau1)
    else:
        g1_pivot = 1.0

    delta_beta0_twist = -twist_shock * g1_pivot
    delta_beta0 = parallel_shift + delta_beta0_twist
    delta_beta1 = slope_shock + twist_shock
    delta_beta2 = curvature1_shock
    delta_beta3 = curvature2_shock

    def get_price_for_params(b0, b1, b2, b3):
        def par_curve_fn(t):
            return nss_yield(t, b0, b1, b2, b3, base_params["tau1"], base_params["tau2"])
        zc = bootstrap_zero_curve(par_curve_fn, max_maturity=40.0, step_size=0.5)
        return calculate_dirty_price(settlement_date, cashflows, zc)

    p_base = get_price_for_params(
        base_params["beta0"],
        base_params["beta1"],
        base_params["beta2"],
        base_params["beta3"]
    )

    shocked_params = apply_scenario_shocks(
        base_params=base_params,
        parallel_shift=parallel_shift,
        slope_shock=slope_shock,
        curvature1_shock=curvature1_shock,
        curvature2_shock=curvature2_shock,
        twist_shock=twist_shock,
        twist_pivot=twist_pivot
    )
    p_shocked = get_price_for_params(
        shocked_params["beta0"],
        shocked_params["beta1"],
        shocked_params["beta2"],
        shocked_params["beta3"]
    )
    total_pnl = (p_shocked - p_base) * (face_value / 100.0)

    # Central difference bump h = 1bp (0.01 percentage point)
    h = 0.01

    p_b0_up = get_price_for_params(base_params["beta0"] + h, base_params["beta1"], base_params["beta2"], base_params["beta3"])
    p_b0_down = get_price_for_params(base_params["beta0"] - h, base_params["beta1"], base_params["beta2"], base_params["beta3"])
    d_b0 = (p_b0_up - p_b0_down) / (2.0 * h)

    p_b1_up = get_price_for_params(base_params["beta0"], base_params["beta1"] + h, base_params["beta2"], base_params["beta3"])
    p_b1_down = get_price_for_params(base_params["beta0"], base_params["beta1"] - h, base_params["beta2"], base_params["beta3"])
    d_b1 = (p_b1_up - p_b1_down) / (2.0 * h)

    p_b2_up = get_price_for_params(base_params["beta0"], base_params["beta1"], base_params["beta2"] + h, base_params["beta3"])
    p_b2_down = get_price_for_params(base_params["beta0"], base_params["beta1"], base_params["beta2"] - h, base_params["beta3"])
    d_b2 = (p_b2_up - p_b2_down) / (2.0 * h)

    p_b3_up = get_price_for_params(base_params["beta0"], base_params["beta1"], base_params["beta2"], base_params["beta3"] + h)
    p_b3_down = get_price_for_params(base_params["beta0"], base_params["beta1"], base_params["beta2"], base_params["beta3"] - h)
    d_b3 = (p_b3_up - p_b3_down) / (2.0 * h)

    contrib_level = d_b0 * delta_beta0 * (face_value / 100.0)
    contrib_slope = d_b1 * delta_beta1 * (face_value / 100.0)
    contrib_curv1 = d_b2 * delta_beta2 * (face_value / 100.0)
    contrib_curv2 = d_b3 * delta_beta3 * (face_value / 100.0)

    residual = total_pnl - (contrib_level + contrib_slope + contrib_curv1 + contrib_curv2)

    return {
        "level": float(contrib_level),
        "slope": float(contrib_slope),
        "curvature1": float(contrib_curv1),
        "curvature2": float(contrib_curv2),
        "residual": float(residual),
        "total": float(total_pnl)
    }

